src/site_apportionment/site_apportionment.py

```diff
@@ -1,98 +1,371 @@
 import pandas as pd
 from pandas.api.types import is_numeric_dtype
-import os
+from typing import Tuple, List, Dict, Callable, Union
 import logging
 
 from src.imputation.imputation_helpers import get_imputation_cols
+from src.outputs.status_filtered import output_status_filtered
 
 SitesApportionmentLogger = logging.getLogger(__name__)
 
-# Colunm names redefined for convenience
-ref = "reference"
-ins = "instance"
-period = "period"
-form = "formtype"
-postcode = "601" # "postcodes_harmonised"
-percent = "602"
-product = "201"
-pg_num = "pg_numeric"
-civdef = "200"
+# Column names redefined for convenience
+ref_col: str = "reference"
+instance_col: str = "instance"
+period_col: str = "period"
+form_col: str = "formtype"
+postcode_col: str = "601"  
+percent_col: str = "602"
+product_col: str = "201"
+pg_num_col: str = "pg_numeric"
+civdef_col: str = "200"
+
+groupby_cols: List[str] = [ref_col, period_col]
+code_cols: List[str] = [product_col, civdef_col, pg_num_col]
```

Comment:
Note to reviewer: Most of the changes here are inherited from branch 660 which we didn't merge.